### PR TITLE
feat(helm_resource): remove CRD manifests

### PR DESCRIPTION
**Pull request by aslafy-z**
Remove installed CRDs on helm_resource Tilt down.

### CODE REVIEW

File: helm_resource/helm-delete-helper.py [hunk 1]
```diff
@@ -19,6 +21,26 @@
   stderr=subprocess.DEVNULL,
 )
 if status_result == 0:
+    # Extract CRDs from managed manifests
+    get_manifest_cmd = ('helm', 'get', 'manifest') + namespace_args + (release_name,)
+    print('Running cmd: %s' % ' '.join(get_manifest_cmd), file=sys.stderr)
+    raw_managed_resources = subprocess.check_output(get_manifest_cmd)
+    managed_resources = list(yaml.load_all(resources, Loader=yaml.SafeLoader))
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:28]: pyyaml is not available. An alternative may be using `grep -Poz '(?s)\s*kind: CustomResourceDefinition.*?metadata:.*?\s+name: \K([-\.a-z0-9]*\n)'`

File: helm_resource/helm-delete-helper.py [hunk 2]
```diff
@@ -6,11 +6,22 @@
...
 
 release_name = os.environ['RELEASE_NAME']
+chart = os.environ['CHART']
 namespace = os.environ.get('NAMESPACE', '')
 namespace_args = ('--namespace', namespace) if namespace else ()
 
+def extract_crd_names(doc):
+  ## As PyYAML is not available, the commented code below is not usable.
+  ## Using a fallback method using regexp.
+  # resources = list(yaml.load_all(resources, Loader=yaml.SafeLoader))
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:19]:
```suggestion
  # resources = list(yaml.load_all(doc, Loader=yaml.SafeLoader))
```

File: helm_resource/helm-delete-helper.py [hunk 3]
```diff
@@ -6,11 +6,22 @@
...
+  # crd_resources = filter(lambda r: r.kind == 'CustomResourceDefinition', resources)
+  # return [r['metadata']['name'] for r in crd_resources]
+  pattern = re.compile(r'\s*kind: CustomResourceDefinition.*?metadata:.*?\s+name: ([-\.a-z0-9]*)\n', re.DOTALL)
+  return list(pattern.findall(doc))
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:23]:
```suggestion
  return list(pattern.findall(docs))
```

File: helm_resource/helm-delete-helper.py [hunk 4]
```diff
@@ -6,11 +6,22 @@
...
 import subprocess
 import sys
+import re
 
 release_name = os.environ['RELEASE_NAME']
+chart = os.environ['CHART']
 namespace = os.environ.get('NAMESPACE', '')
 namespace_args = ('--namespace', namespace) if namespace else ()
 
+def extract_crd_names(doc):
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:16]:
```suggestion
def extract_crd_names(docs):
```

File: helm_resource/helm-delete-helper.py [hunk 5]
```diff
@@ -19,6 +30,25 @@
   stderr=subprocess.DEVNULL,
 )
 if status_result == 0:
+    # Extract CRDs from managed manifests
+    get_manifest_cmd = ('helm', 'get', 'manifest') + namespace_args + (release_name,)
+    print('Running cmd: %s' % ' '.join(get_manifest_cmd), file=sys.stderr)
+    raw_managed_resources = subprocess.check_output(get_manifest_cmd)
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:36]:
```suggestion
    raw_managed_resources = subprocess.check_output(get_manifest_cmd).decode(sys.stdout.encoding)
```

File: helm_resource/helm-delete-helper.py [hunk 6]
```diff
@@ -19,6 +30,25 @@
...
+    crd_managed_names = extract_crd_names(raw_managed_resources)
+  
+    # Extract CRDs from pre-hook installed manifests
+    show_crds_cmd = ('helm', 'show', 'crds') + namespace_args + (release_name,)
+    print('Running cmd: %s' % ' '.join(show_crds_cmd), file=sys.stderr)
+    raw_prehook_crd_resources = subprocess.check_output(get_manifest_cmd)
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:42]:
```suggestion
    raw_prehook_crd_resources = subprocess.check_output(get_manifest_cmd).decode(sys.stdout.encoding)
```

File: helm_resource/helm-delete-helper.py [hunk 7]
```diff
@@ -19,6 +30,25 @@
...
+  
+    # Extract CRDs from pre-hook installed manifests
+    show_crds_cmd = ('helm', 'show', 'crds') + namespace_args + (release_name,)
+    print('Running cmd: %s' % ' '.join(show_crds_cmd), file=sys.stderr)
+    raw_prehook_crd_resources = subprocess.check_output(get_manifest_cmd).decode(sys.stdout.encoding)
+    crd_prehook_names = extract_crd_names(raw_prehook_crd_resources)
+
+    # Remove all discovered CRDs
+    for crd_name in (crd_managed_names + crd_prehook_names):
+        delete_cmd = ('kubectl', 'delete', crd_name,)
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:47]:
```suggestion
        delete_cmd = ('kubectl', 'delete', 'crd', crd_name,)
```

File: helm_resource/test/test.sh [hunk 8]
```diff
@@ -6,3 +6,5 @@ cd "$(dirname "$0")"
 
 tilt ci
 tilt down
+
+kubectl get crd foo && { echo >&2 "crd foo was not cleaned up";  exit 1; }
```
Comment (aslafy-z) [inline helm_resource/test/test.sh:10]:
```suggestion
kubectl get crd foos.helm-resource.tilt-x.dev && { echo >&2 "crd 'foos.helm-resource.tilt-x.dev' was not cleaned up";  exit 1; }
```

File: helm_resource/test/test.sh [hunk 9]
```diff
@@ -6,3 +6,5 @@ cd "$(dirname "$0")"
 
 tilt ci
 tilt down
+
+kubectl get crd foos.helm-resource.tilt-x.dev && { echo >&2 "crd 'foos.helm-resource.tilt-x.dev' was not cleaned up";  exit 1; }
```
Comment (aslafy-z) [inline helm_resource/test/test.sh:10]:
```suggestion
kubectl get crd bars.helm-resource.tilt-x.dev && { echo >&2 "crd 'bars.helm-resource.tilt-x.dev' was not cleaned up";  exit 1; }
kubectl get crd foos.helm-resource.tilt-x.dev && { echo >&2 "crd 'foos.helm-resource.tilt-x.dev' was not cleaned up";  exit 1; }
```

File: helm_resource/helm-delete-helper.py [hunk 10]
```diff
@@ -19,6 +30,25 @@
...
 )
 if status_result == 0:
+    # Extract CRDs from managed manifests
+    get_manifest_cmd = ('helm', 'get', 'manifest') + namespace_args + (release_name,)
+    print('Running cmd: %s' % ' '.join(get_manifest_cmd), file=sys.stderr)
+    raw_managed_resources = subprocess.check_output(get_manifest_cmd).decode(sys.stdout.encoding)
+    crd_managed_names = extract_crd_names(raw_managed_resources)
+  
+    # Extract CRDs from pre-hook installed manifests
+    show_crds_cmd = ('helm', 'show', 'crds') + namespace_args + (release_name,)
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:40]:
```suggestion
    show_crds_cmd = ('helm', 'show', 'crds', chart,)
```

File: helm_resource/helm-delete-helper.py [hunk 11]
```diff
@@ -6,11 +6,22 @@
...
 import sys
+import re
 
 release_name = os.environ['RELEASE_NAME']
+chart = os.environ['CHART']
 namespace = os.environ.get('NAMESPACE', '')
 namespace_args = ('--namespace', namespace) if namespace else ()
 
+def extract_crd_names(docs):
+  ## As PyYAML is not available, the commented code below is not usable.
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:17]:
```suggestion
  print("Will try to extract CRD names from:\n", docs)
  ## As PyYAML is not available, the commented code below is not usable.
```

File: helm_resource/helm-delete-helper.py [hunk 12]
```diff
@@ -6,11 +6,22 @@
...
+  ## Using a fallback method using regexp.
+  # resources = list(yaml.load_all(docs, Loader=yaml.SafeLoader))
+  # crd_resources = filter(lambda r: r.kind == 'CustomResourceDefinition', resources)
+  # return [r['metadata']['name'] for r in crd_resources]
+  pattern = re.compile(r'\s*kind: CustomResourceDefinition.*?metadata:.*?\s+name: ([-\.a-z0-9]*)\n', re.DOTALL)
+  return list(pattern.findall(docs))
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:23]:
```suggestion
  crds = list(pattern.findall(docs))
  print("Found CRDs:", crds)
  return crds
```

File: helm_resource/helm-delete-helper.py [hunk 13]
```diff
@@ -19,6 +33,25 @@
...
+    # Extract CRDs from managed manifests
+    get_manifest_cmd = ('helm', 'get', 'manifest') + namespace_args + (release_name,)
+    print('Running cmd: %s' % ' '.join(get_manifest_cmd), file=sys.stderr)
+    raw_managed_resources = subprocess.check_output(get_manifest_cmd).decode(sys.stdout.encoding)
+    crd_managed_names = extract_crd_names(raw_managed_resources)
+  
+    # Extract CRDs from pre-hook installed manifests
+    show_crds_cmd = ('helm', 'show', 'crds', chart,)
+    print('Running cmd: %s' % ' '.join(show_crds_cmd), file=sys.stderr)
+    raw_prehook_crd_resources = subprocess.check_output(get_manifest_cmd).decode(sys.stdout.encoding)
```
Comment (aslafy-z) [inline helm_resource/helm-delete-helper.py:45]:
```suggestion
    raw_prehook_crd_resources = subprocess.check_output(show_crds_cmd).decode(sys.stdout.encoding)
```

**Comment by nicks** [comment]
ugh, i feel your pain here, i really do.

unfortunately, uninstalling crds is fraught with dragons. you get all sorts of weird bugs where the apiserver removes the API routes associated with the CRD while other stuff still needs to talk to it. There are good reasons why the helm team made this the default `helm uninstall` behavior, and i'm not sure it's this extension's place to override that behavior.

**Pull request closed without merging**